### PR TITLE
executor: do not build range for NullOuterVal in IndexLookUpJoin

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
fix https://github.com/pingcap/tidb/issues/8499

Before this PR, the result of the following query is wrong.
``` sql
CREATE TABLE `t1` (
  `a` int(11) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`)
);
insert into t1 values(1, 0), (2, null);
CREATE TABLE `t2` (
  `a` int(11) NOT NULL,
  PRIMARY KEY (`a`)
);
insert into t2 values(0);
tidb> select /*+ TIDB_INLJ(t2)*/ * from t1 left join t2 on t1.b = t2.a;
+---+------+------+
| a | b    | a    |
+---+------+------+
| 1 |    0 |    0 |
| 1 |    0 |    0 |
| 2 | NULL | NULL |
+---+------+------+
3 rows in set (0.00 sec)
```

### What is changed and how it works?
When the innerPlan is a TableReader, the range of PKHanle will be built in [buildTableReaderForIndexJoin](https://github.com/pingcap/tidb/blob/master/executor/builder.go#L1891-L1901),
and `NilDatum.GetInt64()` will return 0. Thus we read the row with `PkHandle=0` twice, and get a duplicate result line.

When the innerPlan is a IndexReader or IndexLookUpReader, we'll encode the NilDatum, thus no error happens.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects
none

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8505)

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @winoros @eurekaka @lamxTyler

File: executor/builder.go [hunk 1]
```diff
@@ -1895,6 +1895,9 @@ func (builder *dataReaderBuilder) buildTableReaderForIndexJoin(ctx context.Conte
 	}
 	handles := make([]int64, 0, len(datums))
 	for _, datum := range datums {
+		if datum[0].IsNull() {
```
Comment (winoros) [inline executor/builder.go]: Leave a `TODO` here: `After we propagate not null filter on join keys, this check can be removed.

Comment (zz-jason) [reply executor/builder.go]: I prefer to use a github issue to trace this progress.

Comment (zz-jason) [reply executor/builder.go]: @XuHuaiyu After rethinking, maybe it's better to:

In function `buildExecutorForIndexJoin()`, filter out `datums` which contains any `NULL` value. Use the `datums` that doesn't contains any NULL value to build the index key range of the inner table.

**Pull request merged**